src/dht.cpp

```diff
@@ -1341,41 +1398,11 @@ Dht::searchStep(std::shared_ptr<Search> sr)
         while (sent and sr->currentlySolicitedNodeCount() < MAX_REQUESTED_SEARCH_NODES);
         /*DHT_LOG.DEBUG("[search %s IPv%c] step: sent %u requests (total %u).",
             sr->id.toString().c_str(), sr->af == AF_INET ? '4' : '6', i, sr->currentlySolicitedNodeCount());*/
+    }
 
-        auto expiredn = (size_t)std::count_if(sr->nodes.begin(), sr->nodes.end(), [&](const SearchNode& sn) {
-                    return sn.candidate or sn.node->isExpired();
-                });
-        if (i == 0 && expiredn == sr->nodes.size())
-        {
-            DHT_LOG.WARN("[search %s IPv%c] expired", sr->id.toString().c_str(), sr->af == AF_INET ? '4' : '6');
-            // no nodes or all expired nodes
-            sr->expired = true;
-            if (sr->announce.empty() && sr->listeners.empty()) {
-                // Listening or announcing requires keeping the cluster up to date.
-                sr->setDone();
-            }
-            {
-                auto get_cbs = std::move(sr->callbacks);
-                for (const auto& g : get_cbs) {
-                    if (g.second.done_cb)
-                        g.second.done_cb(false, {});
-                }
-            }
-            {
-                std::vector<DoneCallback> a_cbs;
-                a_cbs.reserve(sr->announce.size());
-                for (auto ait = sr->announce.begin() ; ait != sr->announce.end(); ) {
-                    if (ait->callback)
-                        a_cbs.emplace_back(std::move(ait->callback));
-                    if (not ait->permanent)
-                        ait = sr->announce.erase(ait);
-                    else
-                        ait++;
-                }
-                for (const auto& a : a_cbs)
-                    a(false, {});
-            }
-        }
+    if (sr->getNumberOfBadNodes() >= std::min(sr->nodes.size(), static_cast<size_t>(SEARCH_MAX_BAD_NODES))) {
```

Comment:
> - we chose to consider search expiration as a connectivity loss

I might add that it's a connectivity loss with a subset of the network. It's possible we can still reach some nodes. For instance when losing connection only with outside our LAN.

> - another possibility of implementation is to not add a node count limit, and count successive expired nodes, up to Max (our limit for search expiration).

I have implemented this in 05e7ce6. Do you see any reason in favor of implementing the global search limit instead?

> -  But then, what to do with the excessive nodes ? We won't be using them anyway, those nodes are far away, and just before expiring the search we started to send them requests (since they are next in the list after the Max expired nodes).

I don't think it matters. Those nodes will expire and the same process will occur with nodes farther and farther from the search target subsequently used for the search, until all nodes in the cache are expired.

> Yeah it sums it up, but I'm still not sure about how to handle some corner cases (listen recovery...).

Won't the 'listen' operation recover when we are synced again? What do you have in mind?